### PR TITLE
8307483: New micros for j.u.c.LockSupport

### PR DESCRIPTION
**Pull request by ericcaspole**
These micros were developed while investigating JDK-8305670 by myself and Sergey Kuksenko. The order of thread creation was important in that bug, so there are 2 JMH for creating sleepers before and after the worker threads.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8307483](https://bugs.openjdk.org/browse/JDK-8307483): New micros for j.u.c.LockSupport

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to [4a486cfd](https://git.openjdk.org/jdk/pull/13815/files/4a486cfdbb62cff94dc37394ca3278a14143158a)
 * [Claes Redestad](https://openjdk.org/census#redestad) (@cl4es - **Reviewer**) ⚠️ Review applies to [4a486cfd](https://git.openjdk.org/jdk/pull/13815/files/4a486cfdbb62cff94dc37394ca3278a14143158a)

### Contributors
 * Sergey Kuksenko `<skuksenko@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13815/head:pull/13815` \
`$ git checkout pull/13815`

Update a local copy of the PR: \
`$ git checkout pull/13815` \
`$ git pull https://git.openjdk.org/jdk.git pull/13815/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13815`

View PR using the GUI difftool: \
`$ git pr show -t 13815`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13815.diff">https://git.openjdk.org/jdk/pull/13815.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13815#issuecomment-1535366750)

### CODE REVIEW

File: test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    public static class IdleThread implements Runnable {
+        boolean done = false;
+        Thread my_thread;
+
+        @Override
+        public void run() {
+            my_thread = Thread.currentThread();
+            while (!done) {
```
Comment (AlanBateman) [inline test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java:131]: You might want to re-check IdleThread. From a quick look I would have expected "done" to be volatile. Also "my_thread" as it is set in the run with plain access.

Comment (dholmes-ora) [reply test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java:131]: Yes `done` must be volatile. Surprised this even worked as expected.

File: test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+                }
+            });
+        }
+        latch.await();
+    }
+
+    IdleThread[] idle_threads;
+
+
+    ExecutorService exec;
```
Comment (shipilev) [inline test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java:90]:
```suggestion
    IdleThread[] idleThreads;

    ExecutorService exec;
```

File: test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+    @Setup
+    public void setup() throws InterruptedException {
+        barrier = new CyclicBarrier(workers);
+        exec = Executors.newFixedThreadPool(workers);
+        CountDownLatch latch = new CountDownLatch(workers);
+        for (int i = 0; i < workers; i++) { // warmup exec
+            exec.submit(() ->
+            {
```
Comment (shipilev) [inline test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java:99]:
```suggestion
            exec.submit(() -> {
```

File: test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                } finally {
+                    latch.countDown();
+                }
+            });
+        }
+        latch.await();
+        idle_threads = new IdleThread[idles];
+        for(int i=0; i < idle_threads.length; i++) {
+            new Thread(idle_threads[i] = new IdleThread()).start();
+        }
```
Comment (shipilev) [inline test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java:113]: These `idleThreads` are not actually `Thread`-s, they are `Runnable`-s.

```suggestion
        for(int i = 0; i < idles; i++) {
            Runnable r = new IdleRunnable();
            idleRunnables[i] = r;
            new Thread(r).start();
        }
```

File: test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    @TearDown
+    public void tearDown() {
+        for(IdleThread it : idle_threads) {
```
Comment (shipilev) [inline test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java:118]:
```suggestion
        for (IdleThread it : idle_threads) {
```

File: test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+            it.stop();
+        }
+        exec.shutdown();
+    }
+
+    public static class IdleThread implements Runnable {
+        boolean done = false;
+        Thread my_thread;
```
Comment (shipilev) [inline test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java:126]:
```suggestion
        Thread myThread;
```

File: test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java (hunk 1 above)
Comment (shipilev) [reply test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java:131]: Same as we saw with `isAlive` before: `LockSupport.park` implicitly provides a compiler barrier. This `done` should be at least "opaque" to carry the same effect.

**Review by shipilev — COMMENTED** [review]
Cursory style reviews.

File: test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java (hunk 1 above)
Comment (viktorklang-ora) [reply test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java:131]: Yes, `park()` and `unpark()` having release and acquire semantics is not documented [here](https://docs.oracle.com/en/java/javase/20/docs/api/java.base/java/util/concurrent/locks/LockSupport.html#park()), and the documentation for `LockSupport` explicitly says "Reliable usage requires the use of volatile (or atomic) variables to control when to park or unpark. Orderings of calls to these methods are maintained with respect to volatile variable accesses, but not necessarily non-volatile variable accesses."

Comment (dholmes-ora) [reply test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java:131]: Irrespective of park/unpark barriers, the JIT should have hoisted `done` and rewritten as:
```
if (Idone) {
  while(true) {
    ...
  }
}
```

Comment (dholmes-ora) [reply test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java:131]: Well I guess the park/unpark barriers prevented that.

**Comment by ericcaspole** [comment]
I think I fixed all the commented items, could anyone look?

File: test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersBefore.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+    IdleRunnable[] idleRunnables;
+
+    ExecutorService exec;
+
+    @Setup
+    public void setup() {
+        idleRunnables = new IdleRunnable[idles];
+        for(int i = 0; i < idleRunnables.length; i++) {
+            new Thread(idleRunnables[i] = new IdleRunnable()).start();
```
Comment (shipilev) [inline test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersBefore.java:98]:
```suggestion
            idleRunnables[i] = new IdleRunnable();
            new Thread(idleRunnables[i]).start();
```

File: test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersBefore.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        }
+        barrier = new CyclicBarrier(workers);
+        exec = Executors.newFixedThreadPool(workers); // order is important, create this executor only after idle threads
+    }
+
+    @TearDown
+    public void tearDown() {
+        for(IdleRunnable it : idleRunnables) {
+            it.stop();
+        }
```
Comment (shipilev) [inline test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersBefore.java:108]:
```suggestion
        for (IdleRunnable r : idleRunnables) {
            r.stop();
        }
```

File: test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersBefore.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        exec.shutdown();
+    }
+
+    public static class IdleRunnable implements Runnable {
+        volatile boolean done = false;
```
Comment (shipilev) [inline test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersBefore.java:113]:
```suggestion
        volatile boolean done;
```

File: test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    @TearDown
+    public void tearDown() {
+        for (IdleRunnable it : idleRunnables) {
+            it.stop();
+        }
+        exec.shutdown();
+    }
+
+    public static class IdleRunnable implements Runnable {
+        volatile boolean done = false;
```
Comment (shipilev) [inline test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java:125]:
```suggestion
        volatile boolean done;
```

File: test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+            IdleRunnable r = new IdleRunnable();
+            idleRunnables[i] = r;
+            new Thread(r).start();
+        }
+    }
+
+    @TearDown
+    public void tearDown() {
+        for (IdleRunnable it : idleRunnables) {
+            it.stop();
```
Comment (shipilev) [inline test/micro/org/openjdk/bench/java/util/concurrent/UnparkBenchSleepersAfter.java:119]:
```suggestion
        for (IdleRunnable r : idleRunnables) {
            r.stop();
```

**Review by shipilev — APPROVED** [review]
Looks okay, with a few remaining nits.

**Review by cl4es — APPROVED** [review]
LGTM with @shipilev's suggestions

**Comment by ericcaspole** [comment]
/integrate

Thanks Claes and Aleksey!

**Pull request closed without merging**